### PR TITLE
new OutboundEnvelope

### PR DESCRIPTION
**Pull request by patriknw**
- instead of the old Send
- optional recipient, remove of dummy
- pool of OutboundEnvelope

### CODE REVIEW

File: akka-remote/src/main/scala/akka/remote/artery/InboundEnvelope.scala [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+import akka.actor.InternalActorRef
+import akka.util.OptionVal
+import akka.actor.Address
+import akka.actor.ActorRef
+
+/**
+ * INTERNAL API
+ */
+private[akka] object InboundEnvelope {
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/InboundEnvelope.scala]: this is just moved code from ArteryTransport to a separate file

File: akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressionTable.scala [hunk 2]
```diff
@@ -105,7 +103,7 @@ class InboundCompressionTable[T](
     key match {
       case null ⇒ true
       case ""   ⇒ true // empty class manifest for example
-      case _    ⇒ key.endsWith("/system/dummy") || key.endsWith("/") // TODO dummy likely shouldn't exist? can we remove it?
+      case _    ⇒ key.endsWith("/")
```
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressionTable.scala]: great :)

File: akka-bench-jmh/src/main/scala/akka/remote/artery/CodecBenchmark.scala [hunk 3]
```diff
@@ -54,11 +53,9 @@ class CodecBenchmark {
   implicit val system = ActorSystem("CodecBenchmark", config)
   val systemB = ActorSystem("systemB", system.settings.config)
 
-  val envelopePool = new EnvelopeBufferPool(ArteryTransport.MaximumFrameSize, ArteryTransport.MaximumPooledBuffers)
-  val inboundEnvelopePool = new ObjectPool[InboundEnvelope](
-    16,
-    create = () ⇒ new ReusableInboundEnvelope, clear = inEnvelope ⇒ inEnvelope.asInstanceOf[ReusableInboundEnvelope].clear()
-  )
+  private val envelopePool = new EnvelopeBufferPool(ArteryTransport.MaximumFrameSize, ArteryTransport.MaximumPooledBuffers)
+  private val inboundEnvelopePool = ReusableInboundEnvelope.createObjectPool(capacity = 16)
```
Comment (drewhk) [inline akka-bench-jmh/src/main/scala/akka/remote/artery/CodecBenchmark.scala]: is 16 enough?

**Comment by drewhk** [comment]
Any performance improvements by this? What is the cost of using a pool vs. allocating stuff here? With DirectByteBuffers it is worth to keep a pool given how they allocate memory, but is this true for these small objects?

File: akka-bench-jmh/src/main/scala/akka/remote/artery/CodecBenchmark.scala (hunk 3 above)
Comment (drewhk) [inline akka-bench-jmh/src/main/scala/akka/remote/artery/CodecBenchmark.scala]: ah, this is a test...

**Comment by drewhk** [comment]
LGTM

File: akka-bench-jmh/src/main/scala/akka/remote/artery/CodecBenchmark.scala (hunk 3 above)
Comment (patriknw) [inline akka-bench-jmh/src/main/scala/akka/remote/artery/CodecBenchmark.scala]: yes, and there are no async boundaries here, so 1 should be enough

**Comment by patriknw** [comment]
I don't expect any noticeable difference apart from possibly the higher percentiles of the latency.

**Pull request merged**